AEPCore/Tests/RulesTests/EventDataFlattenerTests.swift

```diff
@@ -0,0 +1,213 @@
+            "a": 1,
+            "b": true
+        ]
+        let expected: [String: Any] = [
+            "a": 1,
+            "b": true
+        ]
+        XCTAssertEqualDictionaries(expected, actual.flattening())
+    }
+
+    func testDictionaryWithPrimitives() {
+        let actual: [String: Any] = [
+            "int": 1,
+            "bool": false,
+            "double": 1.23,
+            "string": "a",
+            "null": NSNull()
+        ]
+        let expected: [String: Any] = [
+            "int": 1,
+            "bool": false,
+            "double": 1.23,
+            "string": "a",
+            "null": NSNull()
+        ]
+        XCTAssertEqualDictionaries(expected, actual.flattening())
+    }
+
+    func testNestedDictionaryIsFlattened() {
+        let actual: [String: Any] = [
+            "a": [
+                "b": [
+                    "c": 42
+                ]
+            ]
+        ]
+        let expected: [String: Any] = ["a.b.c": 42]
+        XCTAssertEqualDictionaries(expected, actual.flattening())
+    }
+
+    func testDictionaryWithArrayOfPrimitives() {
+        let actual: [String: Any] = ["nums": [1, true, 1.23, "a", NSNull()]]
+        let expected: [String: Any] = [
+            "nums.0": 1,
+            "nums.1": true,
+            "nums.2": 1.23,
+            "nums.3": "a",
+            "nums.4": NSNull()
+        ]
+        XCTAssertEqualDictionaries(expected, actual.flattening())
+    }
+
+    func testDictionaryWithArrayOfDictionaries() {
+        let actual: [String: Any] = [
+            "arr": [
+                ["x": 1],
+                ["y": 2]
+            ]
+        ]
+        let expected: [String: Any] = [
+            "arr.0.x": 1,
+            "arr.1.y": 2
+        ]
+        XCTAssertEqualDictionaries(expected, actual.flattening())
+    }
+
+    func testMixedNestedStructures() {
+        let actual: [String: Any] = [
+            "dict": [
+                "arr": [
+                    ["dict0": 0],
+                    [
+                        "dict1": 1,
+                        "innerArr": ["a", "b"]
+                    ]
+                ],
+                "bool": false
+            ]
+        ]
+        let expected: [String: Any] = [
+            "dict.arr.0.dict0": 0,
+            "dict.arr.1.dict1": 1,
+            "dict.arr.1.innerArr.0": "a",
+            "dict.arr.1.innerArr.1": "b",
+            "dict.bool": false
+        ]
+        XCTAssertEqualDictionaries(expected, actual.flattening())
+    }
+
+    func testDictionaryNumericalKeys() {
+        let actual: [String: Any] = [
+            "0": [0],
+            "arr": [1]
+        ]
+
+        let expected: [String: Any] = [
+            "0.0": 0,
+            "arr.0": 1
+        ]
+        XCTAssertEqualDictionaries(expected, actual.flattening())
+    }
+
+    func testDictionaryWithDotInKeyIsNotEscaped() {
+        let actual: [String: Any] = [
+            "a.b": ["c": 1],
+            "a": ["b": 2]
```

Comment:
nit: can we also add a case where we have `["a.b" : 1 , "a": ["b": 2]]` and verify the final dictionary has only one key `"a.b"` whose value is `2`